File: n_queens.py
class Solution:
    def __init__(self):
        self.output = []
    def isSafe(self, board, roww, coll):
        n = len(board) - 1
        row = roww
        col = coll
        # checking upward
        while 0 <= row:
            if board[row][col] == 1:
                return False
            row -= 1
        row = roww
        col = coll
        # checking diagonally upward to the left
        while row >= 0 and col >= 0:
            if board[row][col] == 1:
                return False
            row -= 1
            col -= 1
        row = roww
        col = coll
        # checking diagonally upward to the right
        while row >= 0 and col <= n:
            if board[row][col] == 1:
                return False
            row -= 1
            col += 1
        return True
    def solve(self, board, row, col, qpsf, totalQ, result):
        if totalQ == qpsf:
            self.output.append(result)
            return
        for col in range(len(board)):
            if self.isSafe(board, row, col):
                # do
                board[row][col] = 1
                tmp = "." * len(board)
                # recurse
                self.solve(board, row + 1, col, qpsf + 1, totalQ, result +[tmp[:col]+"Q"+tmp[col+1:]])
                # undo
                board[row][col] = 0
        return False

    def solveNQueens(self, n):
        board = [[0 for i in range(n)]for j in range(n)]
        self.solve(board, 0, 0, 0, n, [])
        return self.output

File: n_queens.cpp
//time: O(2^n)
class Solution {
public:
	vector<vector<string>> result;
	bool isSafe(vector<string> &board, int row, int col, int n) {
		int r = row;
		int c = col;
		while (0 <= r) {
			if (board[r][c] == 'Q') {
				return false;
			}
			r--;
		}
		r = row;
		c = col;
		while (r >= 0 and c >= 0) {
			if (board[r][c] == 'Q') {
				return false;
			}
			r--;
			c--;
		}
		r = row;
		c = col;
		while (r >= 0 and c < n) {
			if (board[r][c] == 'Q') {
				return false;
			}
			r--;
			c++;
		}
		return true;
	}
	bool solve(vector<string> &board, int x,  int qpsf, int totalQ) {
		if (totalQ == qpsf) {
			result.push_back(board);
			return true;
		}
		for (int col = 0; col < totalQ; col++) {
			if (isSafe(board, x, col, totalQ)) {
				board[x][col] = 'Q';
				solve(board, x + 1, qpsf + 1, totalQ);
				board[x][col] = '.';
			}
		}
		return false;
	}
	vector<vector<string>> solveNQueens(int n) {
		vector<string> board(n, string(n, '.'));
		solve(board, 0, 0, n);
		return result;

	}
};
